### PR TITLE
Better understandable FAB actions via bottom sheet implementation

### LINKED ISSUE #2415: Better Upload Options
When uploading files or photos using the Nextcloud Android application the options show only icons and no text whatsoever about this. This is rather confusing since the icons are not obvious to their meaning though this is somewhat difficult to achieve with icons in general.

https://i.imgur.com/hEsLcme.png

I propose adding text to the upload such as here:

https://i.imgur.com/q05qwbL.png

This makes life a lot easier and honestly just looks better, in my honest opinion.

### PR DESCRIPTION
**Pull request by AndyScherzinger**
Resolves #2415

* Adds Bottom sheet, replacing the mini FABs
* replaces custom FAB lib with support library version
* PR stacks on top of #2429 for Android Studio 3.1 support
* Adds 3 new lint warnings due to the fact that it is not using compound drawables which interfere with proper R-t-L alignments, so I felt back to not using them...

See themed screenshots
device-2018-04-04-173247
device-2018-04-04-173231

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
cc @mario @tobiasKaminsky @nextcloud/designers for review - corresponding issue is #2415

**Comment by mario** [comment]
UI wise I like this since Talk app uses bottom sheet pretty much everywhere :)

**Comment by AndyScherzinger** [comment]
@mario @tobiasKaminsky beware that due to being stacked on top of #2429 the PR contains many changed that will vanish once #2429 has been merged and this PR has been rebased to master.

**Comment by AndyScherzinger** [comment]
> Looks great! :) Only one thing: The header of the sheet should be »Add to Nextcloud« – or whatever the themed/branded name is.

@jancborchardt Fixed your comment regarding use of "Nextcloud" or whatever the themed/branded name is.

device-2018-04-09-181550

**Comment by AndyScherzinger** [comment]
@mario @tobiasKaminsky please review :-) 
@mario I also took the time to implement the bottom sheet dialog in a separate class using Butterknife :)

**Comment by AndyScherzinger** [comment]
@jancborchardt fixed all your remarks, see https://github.com/nextcloud/android/pull/2432#issuecomment-379809094

@mario @tobiasKaminsky please review :)

**Comment by AndyScherzinger** [comment]
pinging @mario @tobiasKaminsky for dev review :)

**Comment by tobiasKaminsky** [comment]
Tested and working. I will do code review tomorrow, but I do not expect anything serious ;-)
Great enhancement so far :tada:

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetActions.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/*
+ * Nextcloud Android client application
+ *
+ * @author Andy Scherzinger
+ * Copyright (C) 2018 Andy Scherzinger
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * at your option) any later version.
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetActions.java]: The license header looks a bit broken, "at your option)" is maybe a copy & paste error?

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetActions.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+package com.owncloud.android.ui.fragment;
+
+/**
+ * Actions interface to be implemented by any class that makes use of
+ * {@link com.owncloud.android.ui.fragment.OCFileListBottomSheetDialog}.
+ */
+public interface OCFileListBottomSheetActions {
+    /**
+     * creates a folde rwithin the actual folder.
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetActions.java]: typo

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
+/*
+ * Nextcloud Android client application
+ *
+ * @author Andy Scherzinger
+ * Copyright (C) 2018 Andy Scherzinger
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * at your option) any later version.
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java]: Same here

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+    @BindView(R.id.add_to_cloud)
+    TextView mHeadline;
+
+    @BindView(R.id.menu_upload_files)
+    LinearLayout mMenuUploadFiles;
+    @BindView(R.id.menu_upload_from_app)
+    LinearLayout mMenuUploadFromApp;
+    @BindView(R.id.menu_mkdir)
+    LinearLayout mMenuMakeDir;
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java]: We agreed at some point that new classes do not have "m" as prefix for field/member variables.

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+    private Unbinder unbinder;
+
+    private OCFileListBottomSheetActions mActions;
+
+
+    public OCFileListBottomSheetDialog(@NonNull Context context, OCFileListBottomSheetActions
+            actions) {
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java]: can be in one line

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        final View view = getLayoutInflater()
+                .inflate(R.layout.file_list_actions_bottom_sheet_fragment, null);
+        setContentView(view);
+        if (getWindow() != null) {
+            getWindow().setLayout(
+                    ViewGroup.LayoutParams.MATCH_PARENT,
+                    ViewGroup.LayoutParams.WRAP_CONTENT
+            );
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java]: can be in one line

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        }
+
+        unbinder = ButterKnife.bind(this, view);
+        ThemeUtils.tintDrawable(mIconUploadFiles.getDrawable(), ThemeUtils.primaryColor());
+        ThemeUtils.tintDrawable(mIconUploadFromApp.getDrawable(), ThemeUtils.primaryColor());
+        ThemeUtils.tintDrawable(mIconMakeDir.getDrawable(), ThemeUtils.primaryColor());
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java]: Accessing color needs many calls internally, so maybe call it only one time and assign it as a variable.

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+        mHeadline.setText(getContext().getResources().getString(R.string.add_to_cloud,
+                ThemeUtils.getDefaultDisplayNameForRootFolder()));
+
+        mMenuUploadFiles.setOnClickListener(v -> uploadFiles());
+        mMenuUploadFromApp.setOnClickListener(v -> uploadFromApp());
+        mMenuMakeDir.setOnClickListener(v -> createFolder());
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetDialog.java]: Maybe you want to use this? 
The only downside is, that the functions has to be public.

https://github.com/nextcloud/android/blob/c05243b0d3185a09dbc6b4acd8d7e461aa6b1dff/src/main/java/com/owncloud/android/ui/fragment/contactsbackup/ContactsBackupFragment.java#L302-L307

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListFragment.java [hunk 9]
```diff
@@ -327,19 +326,7 @@ public void onActivityCreated(Bundle savedInstanceState) {
             setFabEnabled(false);
         } else {
             setFabEnabled(true);
-            registerFabListeners();
-
-            // detect if a mini FAB has ever been clicked
-            if (PreferenceManager.getFABClicked(getActivity()) > 0) {
```
Comment (tobiasKaminsky) [inline src/main/java/com/owncloud/android/ui/fragment/OCFileListFragment.java]: getFABClicked() and setFABClicked() are not removed in PreferencesManager.

File: src/main/res/drawable/ic_plus.xml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<!-- drawable/plus.xml -->
```
Comment (tobiasKaminsky) [inline src/main/res/drawable/ic_plus.xml]: Please specify a license header

File: src/main/res/layout/account_action.xml [hunk 11]
```diff
@@ -19,12 +19,13 @@
...
-<LinearLayout
-    xmlns:android="http://schemas.android.com/apk/res/android"
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
     android:layout_width="fill_parent"
     android:layout_height="@dimen/account_action_layout_height"
     android:orientation="horizontal"
-    android:weightSum="1">
+    android:weightSum="1"
+    tools:ignore="UseCompoundDrawables">
```
Comment (tobiasKaminsky) [inline src/main/res/layout/account_action.xml]: This is independent of this PR, or?
If so, please remove it and also the other for the same type.

(on a side note: if this compound drawable lint warning is not working for us, we can think of disabling it)

File: src/main/res/layout/file_list_actions_bottom_sheet_fragment.xml [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+        android:paddingLeft="@dimen/standard_padding"
+        android:paddingRight="@dimen/standard_padding"
+        android:paddingTop="@dimen/standard_half_padding"
+        tools:ignore="UseCompoundDrawables">
+
+        <ImageView
+            android:id="@+id/menu_icon_upload_from_app"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:contentDescription="@null"
```
Comment (tobiasKaminsky) [inline src/main/res/layout/file_list_actions_bottom_sheet_fragment.xml]: This is set to null because of the textView next to it?

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger only minor CR changes :tada:
If you do not have the time, I can handle them for you.

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky if you have the time, please feel free, rather busy work week atm :(

File: src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetActions.java (hunk 1 above)
Comment (tobiasKaminsky) [reply src/main/java/com/owncloud/android/ui/fragment/OCFileListBottomSheetActions.java]: https://www.gnu.org/licenses/gpl-3.0.de.html 
according to this (at your option) is the new part, so I will change this and then will open up a new PR fixing the other files where "(" is missing.

File: src/main/res/layout/file_list_actions_bottom_sheet_fragment.xml (hunk 12 above)
Comment (tobiasKaminsky) [reply src/main/res/layout/file_list_actions_bottom_sheet_fragment.xml]: Good point :+1:

**Comment by tobiasKaminsky** [comment]
@AndyScherzinger done, please have a quick review of my changes :-)

File: src/main/res/drawable/ic_plus.xml [hunk 13]
```diff
@@ -1,4 +1,20 @@
-<!-- drawable/plus.xml -->
+<!--
+    @author Andy Scherzinger
+    Copyright (C) 2018 Andy Scherzinger
```
Comment (AndyScherzinger) [inline src/main/res/drawable/ic_plus.xml]: The author is Google, not me ;)

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky all fine by me except for the icon copyright which is Google, not me :) :+1:

**Comment by AndyScherzinger** [comment]
:+1: all good :)

**Comment by tobiasKaminsky** [comment]
:+1:

**Comment by AndyScherzinger** [comment]
Cc @mario for final approval :)

**Comment by mario** [comment]
👍

**Pull request merged**